### ISSUE TITLE
Regression - predicting shape, size, and position

### ISSUE THREAD

**Issue by nikogamulin** [nicolas-chaulet/torch-points3d#502]
Hi,

I have created a custom dataset of point clouds representing [superquadrics](https://en.wikipedia.org/wiki/Superquadrics). I would like to predict the superquadric's shape (epsilon parameters), size (x, y, and z dimension) and offset (vector from point (0, 0, 0) to the center of the object).

In order to do that, I have modified the architecture of RSConv in order to have 8 outputs, and used MSE as a loss function as this is a regression problem.

The epsilon value predictions are quite good, the size predictions are not that good and the offset value predictions are way too bad. Below I am pasting the charts of loss values:

![loss_epsilon](https://user-images.githubusercontent.com/1803701/99898185-d2069100-2c9f-11eb-98b3-93351c1d7778.png)
![loss_dimension](https://user-images.githubusercontent.com/1803701/99898190-d6cb4500-2c9f-11eb-905c-dc4d74a3cc2b.png)
![loss_offset](https://user-images.githubusercontent.com/1803701/99898182-c87d2900-2c9f-11eb-9a52-e822f5ba865c.png)

Could anyone comment if such predictions are expected - is the architecture capable of detecting a position or is it good only to predict shape and size. I had similar problems with PointNet and I believe that architecture is not good for determining the shape location in space but as far as RSConv is concerned, I believe it should be possible to determine the position as well.

Another question/concern regarding the loss values (images above) are the spikes - I don't know why around step 100 the loss values suddenly increase and then drop again. I would be very thankful if anyone could comment.

I have seen that some other people (@ttsesm) are directed toward a similar goal. As in the end, I am planning to publish an article, I would be glad to connect with other people working on similar issues so we could help each other out and reach our goals faster.

**Comment by ttsesm** [comment]
Hi @nikogamulin, it seems an interesting project. Well I am no sure how well enough I can help, since my experience is kind of limited and I also just got into the field. 

In any case some questions that I have, and I guess it might also help someone else (possibly @nicolas-chaulet) to provide any kind of feedback are the following:

1. Are your input point clouds (x, y, z points) dense or sparse?
2. How do you feed the parameters into the network, I guess you pass the epsilon values and the offset as a feature vector `x` in the data object. As I see it you need to pass the x, y, z there as well right? Except from providing them as `pos`?
3. How do you get the losses? I guess you do it individually for each feature vector that you are regressing and you want to predict.
4. Did you try different batch sizes, from what I've read/heard around this could possibly make some difference.

Thanks.

**Comment by nikogamulin** [comment]
@ttsesm , [here](https://github.com/nikogamulin/superquadrics/blob/main/custom-dataloader-v2.ipynb) is my notebook where custom dataloader and dataset is defined. Hope it helps.

**Comment by nicolas-chaulet** [comment]
Hey @nikogamulin the test logs are a bit suspicious, you should not see those huge spikes, it means that for those particular epochs the network was doing really really bad on your test set. Also, hard to comment since those just cancel any information that we could read from the graphs due to the squashed scale. Do you have a wandb version of those logs?

**Comment by ttsesm** [comment]
@nikogamulin but are you doing regression in the example that you linked, because looking in the code my understanding is that you are still doing classification. 

I might be wrong though :$.

**Comment by ngamulin** [comment]
@ttsesm , In the linked example, I am doing classification. I haven't published a regression sample yet.

**Comment by nikogamulin** [comment]
Here are some example comparisons of point clouds generated with predicted parameters (shape, dimension, offset) - red color, against the point clouds generated with ground truth parameters - blue:

![Figure_12](https://user-images.githubusercontent.com/1803701/100107762-4ee56680-2e6a-11eb-92e1-3452c4ed5374.png)
![Figure_11](https://user-images.githubusercontent.com/1803701/100107766-4f7dfd00-2e6a-11eb-9d37-fed27e622b87.png)
![Figure_10](https://user-images.githubusercontent.com/1803701/100107767-50169380-2e6a-11eb-901f-92b9ab246154.png)
![Figure_9](https://user-images.githubusercontent.com/1803701/100107770-50169380-2e6a-11eb-9842-ec29dc0d5a0e.png)
![Figure_8](https://user-images.githubusercontent.com/1803701/100107771-50af2a00-2e6a-11eb-90d5-bbd99b96bbe9.png)
![Figure_5](https://user-images.githubusercontent.com/1803701/100107773-50af2a00-2e6a-11eb-8031-5bbc36f6e4dc.png)

The shape parameters are predicted quite well (in some cases it seems that it is not the case as some objects are squashed due to dimension error).

I still haven't found out what causes the error spikes on the above charts (previous posts). Anyway, thanks @nicolas-chaulet for fast response. Hope to improve the results soon.

**Comment by nikogamulin** [comment]
> @ttsesm , In the linked example, I am doing classification. I haven't published a regression sample yet.

[Here](https://github.com/nikogamulin/superquadrics/blob/main/superquadrics.py) is the regression (and classification) script

**Comment by ttsesm** [comment]
@nikogamulin one question why is the `input_nc = 3` in your regressor? what are the features that you pass in?

**Comment by nikogamulin** [comment]
With modification of transformation options, I have managed to achieve more meaningful results, Here are some images of ground-truth pointcouds and pointclouds, generated by using predicted parameters:
![Figure_2](https://user-images.githubusercontent.com/1803701/101484577-237a7580-395a-11eb-8198-08e0f2b1bf31.png)
![Figure_5](https://user-images.githubusercontent.com/1803701/101484604-2d9c7400-395a-11eb-8697-395e869b7a99.png)
![Figure_1](https://user-images.githubusercontent.com/1803701/101484609-30976480-395a-11eb-9a6f-a0e397dc7cb0.png)

I have forked the repo and added the scripts for superquadrics regression: https://github.com/nikogamulin/torch-points3d

**Comment by nicolas-chaulet** [comment]
Looks great!

Le mar. 8 déc. 2020 à 12:36, Niko Gamulin <notifications@github.com> a
écrit :

**Comment by nikogamulin** [comment]
I believe this issue can be closed
